### ISSUE TITLE
Generic Response Types

### ISSUE THREAD

**Issue by croman17** [swagger-api/swagger-core#3323]
When an API returns a generic response type swagger-core:
1. The swagger-core generated object name  is not linked to the API definition response.  
2. Does not provide a means to customize the name of the returned object.   

Example:   
ResponseBody<List<String>>  getTeamMemberNames(Long teamId);  

Swagger-core generates:  
schema object:   ResponseBodyListString   (not a name we would want to expose)

      "ResponseBodyListString" : {
        "type" : "object",
        "properties" : {
          "data" : {
            "type" : "array",
            "items" : {
              "type" : "string"
            }
          }
        }
      },

API definition: 
    "/team/membernames" : {
      "get" : {
        "operationId" : "getTeamMemberNames",
        "parameters" : [ {
          "name" : "teamId",
          "in" : "path",
          "required" : true,
          "schema" : {
            "type" : "integer",
            "format" : "int64"
          }
        } ],
        "responses" : {
          "200" : {
            "description" : "OK",
            "content" : {
              "application/json" : {
                "schema" : {
                  "$ref" : "#/components/schemas/ResponseBody"
                }
              }
            }
          },
        }
      }
    },
 
You can see above, that the  API definition response returns  "ResponseBody"  the name of the generic class.   Rather than the generated  ResponseBodyListString  object.    

Current workaround:  We define a custom "doc" class and  use that in the Operation annotation.  
This requires  the doc class and the actual implementation are kept in-sync.   

    @Operation(summary = "Gets team member names", 
        description = "Gets list of team member names.",
        responses = {
             @ApiResponse(responseCode = "200", description = "OK", 
                 content = @Content(schema = @Schema(implementation = TeamMemberNamesResponse.class))),
         })

**Comment by idkw** [comment]
This is really bad ...
Generics are not supported properly due to this issue.  
How can we help ?

**Comment by amjadaziz817** [comment]
@idkw @croman17 I am not if you were able to resolve the issue or still facing this issue. I was also stuck with this scenario and doing some POCs. I was able to handle this scenario by not providing `content` definition in the ApiResponse. 
`@Operation(summary = "Gets team member names",
  description = "Gets list of team member names.",
    responses = {
         @ApiResponse(responseCode = "200", description = "OK")
     })`
     
Can you try above definition and it should work for generic responses.

**Comment by chelsternp44** [comment]
I am seeing the similar issue.
**Controller**
```
@GetMapping( "/loads" )
public LoadPage getShipmentLoads
```

```
public class LoadPage extends Page<Load>
```

```
public class Page<T> {
     String one; //able to see in swagger
     ObjectClass foo; //able to see in swagger
     
     List<T> realInfo; //not able to see in swagger
}
```
`Load` class is simple POJO.

If I follow advice I've seen other places by creating a private class in the controller as such it works, but this is not a great solution since `Page` is in a separate library.  
```
private static class SwaggerLoadPage {

        String one;
        ObjectClass foo;

        List<Load> results;
    }
```

We are currently trying to migrate from Springfox to OpenApi.  This somehow works with SpringFox as is.  I think this could also be tied to the fact that `Page` is in a different library and using different versions of Swagger....

**Comment by idkw** [comment]
I found the issue on my side, it appeared that if you annotate you class with a @Schema and a name springfox doesn't generate a distinct schema for every variant, but if you don't annotate it, it does generate a definition for every variant.

TL; DR; : 

Don't do this :

```
@Schema(name="Page")
public class ListResponse<T> {     
     List<T> data;
     // ... other fields
}
```

But do this : 

```
public class ListResponse<T> {     
     List<T> data; 
     // ... other fields
}
```

**Comment by chelsternp44** [comment]
unfortunately our generic field is nested in a class from a separate library that I don't think we can make changes to easily.

Also we were on SpringFox it was working but with `springdoc-openapi` it is not

**Comment by idkw** [comment]
@chelsternp44 while you work with the maintainer of the separate library to remove the @Schema annotation, you can remove it temporary using bytecode manipulation on your side at runtime, have a look at https://uzxmx.github.io/add-or-remove-java-annotation-at-runtime.html

This is really hacky but at least it works

**Comment by chelsternp44** [comment]
It turns out that our issue is due to a javax annotation breaking it.  We are no longer seeing the issue once we dealt with that.
```
import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;

...
@XmlAccessorType(XmlAccessType.NONE)
```

**Comment by varunsufi** [comment]
@chelsternp44  Is there any update on this issue ??

**Comment by idkw** [comment]
> @chelsternp44 Is there any update on this issue ??

Have you tried this workaround, it has been working great for us so far

https://github.com/swagger-api/swagger-core/issues/3323#issuecomment-917043167

**Comment by seabird86** [comment]
> [#3323 (comment)](https://github.com/swagger-api/swagger-core/issues/3323#issuecomment-917043167)
This is a workaround, but it's still so bad. More code, more class and definition and have to sync them.
Is this any plan and solution to fix it ? 
@croman17 @idkw 

- _The swagger-core generated object name is not linked to the API definition response._ ->  Why does api definition only has ResponseBody ? why don't generate with ResponseBody$List$String. Because of swagger core ? I remember springfox can do it.

- _Does not provide a means to customize the name of the returned object._ -> new version of spring doc has @Content(@schema). However It can't write (implement= ResponseBody<List<String>> .class) in annotation.

**Comment by seabird86** [comment]
`text/plain:
              schema:
                $ref: '#/components/schemas/Response<Money>'
components:
  schemas:
    Response<Money>:
      type: object
      properties:
        data:
          $ref: '#/components/schemas/Money'
    Money:
      type: object
      properties:
        amount:
          type: integer
          format: int32`
I see in micronaut, it happen like this

**Comment by seabird86** [comment]
Awesome, I can customize with generic response: 

```
@Bean
    public OperationCustomizer customize(){
        return (operation,method) ->{
            ApiResponses responses = operation.getResponses();
            if (method.getMethod().getReturnType().equals(ResponseEntity.class)){
                Type type = ((ParameterizedType)((ParameterizedType)method.getMethod().getGenericReturnType()).getActualTypeArguments()[0]).getActualTypeArguments()[0];
                ResolvedSchema resolvedSchema = ModelConverters.getInstance().resolveAsResolvedSchema(new AnnotatedType(type));
                Map<String, Schema> schemas = openAPIService.getCalculatedOpenAPI().getComponents().getSchemas();
                schemas.put(resolvedSchema.schema.getName(),resolvedSchema.schema);
                Schema schema = new ObjectSchema()
                        .type("object")
                        .addProperty("data",schemas.get(resolvedSchema.schema.getName()))
                        .name("ResponseBody<%s>".formatted(resolvedSchema.schema.getName()));
                schemas.put("ResponseBody<%s>".formatted(resolvedSchema.schema.getName()),schema);
                responses.addApiResponse("Success",new ApiResponse().content(new Content().addMediaType("application/json",new MediaType().schema(new ObjectSchema().$ref(schema.getName())))));
            }

            return operation;
        };
    }
```

**Comment by santosfyuri** [comment]
@seabird86 Thanks a lot!

**Comment by shadon178** [comment]
> I found the issue on my side, it appeared that if you annotate you class with a @Schema and a name springfox doesn't generate a distinct definition for every variant, but if you don't annotate it, it does generate a definition for every variant. Not sure if this is a bug though ?
> 
> TL; DR; :
> 
> Don't do this :
> 
> ```
> @Schema(name="Page")
> public class ListResponse<T> {     
>      List<T> data;
>      // ... other fields
> }
> ```
> 
> But do this :
> 
> ```
> public class ListResponse<T> {     
>      List<T> data; 
>      // ... other fields
> }
> ```
> 
> Then when you have a controller with something like :
> 
> ```
> public ListResponse<Book> getBooks() {
>    return new ListResponse<>(List.of(new Book("1")));
> }
> ```
> 
> Then you'll get a proper definition of something like ListResponseBook

I have the same question.

**Comment by zhangk-info** [comment]
must **extends ResponseEntity<T>**
 
 public class ResponseSupportConverter implements ModelConverter {

	/**
	 * The Spring doc object mapper.
	 */
	private final ObjectMapperProvider springDocObjectMapper;

	/**
	 * Instantiates a new Response support converter.
	 *
	 * @param springDocObjectMapper the spring doc object mapper
	 */
	public ResponseSupportConverter(ObjectMapperProvider springDocObjectMapper) {
		this.springDocObjectMapper = springDocObjectMapper;
	}

	@Override
	public Schema resolve(AnnotatedType type, ModelConverterContext context, Iterator<ModelConverter> chain) {
		JavaType javaType = springDocObjectMapper.jsonMapper().constructType(type.getType());
		if (javaType != null) {
			Class<?> cls = javaType.getRawClass();
			if (**isResponseTypeWrapper(cls)** && !isFluxTypeWrapper(cls)) {
				JavaType innerType = javaType.getBindings().getBoundType(0);
				if (innerType == null)
					return new StringSchema();
				return context.resolve(new AnnotatedType(innerType)
						.jsonViewAnnotation(type.getJsonViewAnnotation())
						.ctxAnnotations((type.getCtxAnnotations()))
						.resolveAsRef(true));
			}
			else if (isResponseTypeToIgnore(cls))
				return null;
		}
		return (chain.hasNext()) ? chain.next().resolve(type, context, chain) : null;
	}

}

**Comment by FranH20** [comment]
I made this, Created a class with generic type named PaginationResponse, that I wanted to use as the response type for swagger.
I defined an extra class, named SwaggerCategoryResponse that extends PaginationResponse and it seems to work as expected now.

`public class SwaggerCategoryResponse extends PaginationResponse<CategoryResponse>`


![Imagen2](https://user-images.githubusercontent.com/54047442/235055984-dbeedce8-e452-4bff-8589-030c62ce3b71.png)
![Imagen1](https://user-images.githubusercontent.com/54047442/235055990-94788d21-08b3-46be-abb0-45042e59bff7.png)

**Comment by seabird86** [comment]
Micronaut nailed it. I think, because of limitation of Annotation for java, so Spring-doc cannot do, and wait some one impove spring-doc again.

**Comment by 924060929** [comment]
Thank you for your inspiration @seabird86 .
 In springdoc 1.7.0 (with swagger-models-2.2.9), I change the code for supply the generic type support

my generic type java bean is
```java
@Data
@NoArgsConstructor
@AllArgsConstructor
public class Resp<T> {
    @Schema(description = "0: ok; 1: failed")
    private int code;

    @Schema(description = "ok or failed")
    private String msg;

    @Schema(description = "data", nullable = true)
    private Object data;
}
```

and I add this config to support generic type
```java
@Configuration
public class OpenApiConfig {

    private Map<String, Schema> respSchemas = new ConcurrentHashMap<>();

    @Bean
    public OpenApiCustomiser addGenericSchemaToOpenApi() {
        return openApi -> {
            Map<String, Schema> schemas = openApi.getComponents().getSchemas();

            Set<String> shouldRemoveRespSchemas = schemas.entrySet()
                    .stream()
                    .filter(kv -> {
                        if (!kv.getKey().startsWith(Resp.class.getSimpleName())) {
                            return false;
                        }
                        Map fieldsProperties = kv.getValue().getProperties();
                        Field[] respFields = Resp.class.getDeclaredFields();
                        if (fieldsProperties.size() != respFields.length) {
                            return false;
                        }

                        Set<String> respFieldSet = Arrays.stream(respFields)
                                .map(Field::getName)
                                .collect(Collectors.toSet());

                        if (Sets.intersection(fieldsProperties.keySet(), respFieldSet).size() != respFields.length) {
                            return false;
                        }
                        return true;
                    })
                    .map(Entry::getKey)
                    .collect(Collectors.toSet());

            shouldRemoveRespSchemas.forEach(schemas::remove);
            schemas.putAll(respSchemas);
        };
    }

    @Bean
    public OperationCustomizer customize() {
        // 1. collect RespXxx
        // 2. replace $ref RespXxx to Resp<Xxx>

        return (operation,method) -> {
            ApiResponses responses = operation.getResponses();
            if (method.getMethod().getReturnType().equals(Resp.class)) {

                ResolvedSchema baseRespSchema = ModelConverters.getInstance()
                        .resolveAsResolvedSchema(new AnnotatedType(Resp.class));

                Map<String, Schema> fieldsSchema = Maps.newLinkedHashMap();
                fieldsSchema.putAll(baseRespSchema.schema.getProperties());

                Class actualTypeArgument = (Class) ((ParameterizedTypeImpl) method.getMethod()
                        .getGenericReturnType()).getActualTypeArguments()[0];
                ResolvedSchema resolvedSchema = ModelConverters.getInstance()
                        .resolveAsResolvedSchema(new AnnotatedType(actualTypeArgument));
                String respSchemaName;
                if (resolvedSchema.schema != null) {
                    // override data field schema
                    if (resolvedSchema.referencedSchemas.isEmpty()) {
                        fieldsSchema.put("data", resolvedSchema.schema);
                    } else {
                        fieldsSchema.put("data", new ObjectSchema().$ref(actualTypeArgument.getSimpleName()));
                    }

                    respSchemas.putAll(resolvedSchema.referencedSchemas);

                    respSchemaName = "Resp<" + actualTypeArgument.getSimpleName() + ">";
                } else {
                    // override data field schema
                    Schema originDataSchema = fieldsSchema.get("data");
                    fieldsSchema.put("data", new MapSchema()
                            .description(originDataSchema.getDescription())
                            .nullable(originDataSchema.getNullable()));

                    respSchemaName = "Resp<Map>";
                }

                Schema schema = new ObjectSchema().type("object")
                        .properties(fieldsSchema)
                        .name(respSchemaName);

                // // replace ref '#/components/schemas/RespXxx' to '#/components/schemas/Resp<Xxx>'
                for (ApiResponse apiResponse : responses.values()) {
                    for (MediaType mediaType : apiResponse.getContent().values()) {
                        Schema originApiResponseSchema = mediaType.getSchema();
                        if (originApiResponseSchema.get$ref() != null
                                && originApiResponseSchema.get$ref().startsWith("#/components/schemas/Resp")) {
                            originApiResponseSchema.$ref(schema.getName());
                        }
                    }
                }

                respSchemas.put(respSchemaName, schema);
            }

            return operation;
        };
    }
}
```

And finally swagger-ui can identify the generic / nested structure:
![image](https://github.com/swagger-api/swagger-core/assets/8806055/c13d3bf9-540f-42b8-a3af-0c0407fb89f0)

**Comment by RodionKorneev** [comment]
If you wanna just to change the default name generation behaviour to generic types, you could replace default `TypeNameResolver` with your custom implementation. 
You need to override method 
```java
protected String nameForGenericType(JavaType type, Set<Options> options)
```
with your logic.

For example:
```java
import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.apache.commons.text.WordUtils;

import com.fasterxml.jackson.databind.JavaType;

import io.swagger.v3.core.jackson.TypeNameResolver;
import io.swagger.v3.core.util.PrimitiveType;

public class CustomTypeNameResolver extends TypeNameResolver {

    @Override
    protected String nameForGenericType(JavaType type, Set<Options> options) {
        String baseName = this.nameForClass(type, options);
        int count = type.containedTypeCount();

        List<String> genericNames = new ArrayList<>();
        for (int i = 0; i < count; ++i) {
            JavaType arg = type.containedType(i);
            String argName = PrimitiveType.fromType(arg) != null ? this.nameForClass(arg, options)
                    : this.nameForType(arg, options);
            genericNames.add(WordUtils.capitalize(argName));
        }
        String generic = genericNames.stream().collect(Collectors.joining(", ", "<", ">"));
        return baseName + generic; // will return "Example<Foo, Bar>" 
    }
}
```

**Comment by taku-hua** [comment]
i made it just using the annotation `@Schema`  at my generic data  .my  commonResult bean is 

```java
@Slf4j
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
@Schema(description = "the common response")
public class CommonResult<T> extends ErrorInfo implements Serializable{

   
    @Schema(description = "responseCode", type = "Integer")
    @NotNull
    private Integer code;
    
    @Schema(description = "responseMsg", type = "String")
    private String message;
    //  using the subTypes  to support generic response type
    @Schema(description  = "response data",subTypes = Objects.class)

    private T data;

    @Schema(description = "request id", type = "String")
    private String requestId;


}
```

and my dependency is springfox 3.0.0
```xml
          <dependency>
                <groupId>io.springfox</groupId>
                <artifactId>springfox-boot-starter</artifactId>
                <version>3.0.0</version>
            </dependency>
```
And finally swagger-ui can identify the generic / nested structure:
![image](https://github.com/swagger-api/swagger-core/assets/59927454/ba4f52e5-4bc9-43ba-bf98-be7981361c72)

**Comment by ytk929** [comment]
> i made it just using the annotation `@Schema` at my generic data .my commonResult bean is
> 
> ```java
> @Slf4j
> @Data
> @Builder
> @NoArgsConstructor
> @AllArgsConstructor
> @EqualsAndHashCode(callSuper = true)
> @Schema(description = "the common response")
> public class CommonResult<T> extends ErrorInfo implements Serializable{
> 
>    
>     @Schema(description = "responseCode", type = "Integer")
>     @NotNull
>     private Integer code;
>     
>     @Schema(description = "responseMsg", type = "String")
>     private String message;
>     //  using the subTypes  to support generic response type
>     @Schema(description  = "response data",subTypes = Objects.class)
> 
>     private T data;
> 
>     @Schema(description = "request id", type = "String")
>     private String requestId;
> 
> 
> }
> ```
> 
> and my dependency is springfox 3.0.0
> 
> ```
>           <dependency>
>                 <groupId>io.springfox</groupId>
>                 <artifactId>springfox-boot-starter</artifactId>
>                 <version>3.0.0</version>
>             </dependency>
> ```
> 
> And finally swagger-ui can identify the generic / nested structure: ![image](https://user-images.githubusercontent.com/59927454/269616153-ba4f52e5-4bc9-43ba-bf98-be7981361c72.png)

It solved my problem.

**Comment by cloudlessa** [comment]
Is there an official solution to this problem? 
don't use a custom method to implementing like this:
![image](https://github.com/swagger-api/swagger-core/assets/20315613/48abdc3d-b964-4618-8365-9ed35e9bc0c8)

Springdoc directly calls the method inside swagger-core, but Springfox has its own method of parsing the paradigm.
```
public class DefaultGenericTypeNamingStrategy implements GenericTypeNamingStrategy {
  private static final String OPEN = "«";
  private static final String CLOSE = "»";
  private static final String DELIM = ",";
  //other code xxxx
}
```

**Comment by atrifyllis** [comment]
Why would I want ResponseBody<List<String>> to be mapped to ResponseBodyListString in the first place? This does not make sense, I don't want a new type to be generated. I want ResponseBody to be generated and then probably items.type should contain all possible types found in the endpoints? (in this case only string). Is there any way to do that?

**Comment by AlexElin** [comment]
There's a blog post on this topic: https://json-schema.org/blog/posts/dynamicref-and-generics

Maybe it will be helpful

**Comment by formingform** [comment]
> ```java
> Class actualTypeArgument = (Class) ((ParameterizedTypeImpl) method.getMethod()
>                         .getGenericReturnType()).getActualTypeArguments()[0];
> ```

at this line:
```
Class actualTypeArgument = (Class) ((ParameterizedTypeImpl) method.getMethod()
                        .getGenericReturnType()).getActualTypeArguments()[0];
```
I got a error:
```
Inconvertible types; cannot cast 'java.lang.reflect.Method' to 'com.alibaba.fastjson.util.ParameterizedTypeImpl'
```
what's wrong?

**Comment by seabird86** [comment]
@formingform   'method' is variable in the interface OperationCustomizer. You open this interface and check the class of this variable. Not sure why you are using reflect.Method

**Comment by ochnios** [comment]
> I found the issue on my side, it appeared that if you annotate you class with a @Schema and a name springfox doesn't generate a distinct definition for every variant, but if you don't annotate it, it does generate a definition for every variant. Not sure if this is a bug though ?
> 
> TL; DR; :
> 
> Don't do this :
> 
> ```
> @Schema(name="Page")
> public class ListResponse<T> {     
>      List<T> data;
>      // ... other fields
> }
> ```
> 
> But do this :
> 
> ```
> public class ListResponse<T> {     
>      List<T> data; 
>      // ... other fields
> }
> ```
> 
> Then when you have a controller with something like :
> 
> ```
> public ListResponse<Book> getBooks() {
>    return new ListResponse<>(List.of(new Book("1")));
> }
> ```
> 
> Then you'll get a proper definition of something like ListResponseBook

Yeah, it works! However, it is important to use List<T> - it doesn't work with just Iterable<T> for some reason.

**Comment by handrews** [comment]
@AlexElin yes, that blog post is the correct solution, as endorsed by the [OpenAPI Specification](https://spec.openapis.org/oas/v3.1.1#generic-template-data-structures).  The problem appears to be that Swagger still does not support `$dynamicRef` and `$dynamicAnchor` (see #4845).

**Comment by MatthiasGrandl** [comment]
Stupid question, but now that this is in swagger, how do you actually use it?

I tried this, but it doesn't seem to do anything:
```kt
@Schema(description = "A paginated result.", `$dynamicAnchor` = "T")
data class PaginationResponse<T>(
	@field:Schema(
		description = "List of items on this page. May be less than the requested limit if not enough items are available.",
		`$dynamicRef` = "T"
	)
	val data: List<T>,
	@field:Schema(description = "Total items available under the requested filter.")
	val total: Int,
	@field:Schema(description = "Offset in the list of total items that was requested.")
	val after: Int,
	@field:Schema(description = "Number of items requested.")
	val limit: Int?,
)
```

**Comment by bsayli** [comment]
**Practical workaround + proposal for generic response wrappers**

We hit the same limitation when documenting a generic envelope like `ServiceResponse<T>` and then generating clients. Tooling tends to invent names such as `ServiceResponseListString` (or similar), while what we *really* need downstream is a thin, type-safe wrapper that preserves the generic parameter.

###  ✅ What works well in practice (Spring Boot + Springdoc + OpenAPI Generator)

**Producer side (Springdoc):** during OpenAPI build we detect `T` from controller return types and register a concrete schema per payload *with vendor extensions* that label it as a wrapper, e.g.:

```yaml
components:
  schemas:
    ServiceResponseCustomerCreateResponse:
      allOf:
        - $ref: "#/components/schemas/ServiceResponse"
        - type: object
          properties:
            data:
              $ref: "#/components/schemas/CustomerCreateResponse"
      x-api-wrapper: true
      x-api-wrapper-datatype: CustomerCreateResponse
```

**Consumer side (OpenAPI Generator):** a tiny Mustache overlay checks those extensions and emits a thin shell *instead of* duplicating envelope fields:

```mustache
// api_wrapper.mustache
import {{commonPackage}}.ServiceClientResponse;

public class {{classname}}
    extends ServiceClientResponse<{{vendorExtensions.x-api-wrapper-datatype}}> {
}
```

…and `model.mustache` delegates to that partial only when the wrapper hint is present:

```mustache
{{#vendorExtensions.x-api-wrapper}}
  {{>api_wrapper}}
{{/vendorExtensions.x-api-wrapper}}
{{^vendorExtensions.x-api-wrapper}}
  {{#isEnum}}{{>modelEnum}}{{/isEnum}}
  {{^isEnum}}{{>pojo}}{{/isEnum}}
{{/vendorExtensions.x-api-wrapper}}
```

### 🎯 Result

```java
public class ServiceResponseCustomerCreateResponse
        extends ServiceClientResponse<CustomerCreateResponse> {}
```

No duplicated `status/message/errors` per endpoint; `getData()` is strongly typed.

### Why posting this here

This issue (#3323) is about the gap between how swagger-core names generic schemas and how API producers/consumers expect to use them.  

Related: on the OpenAPI Generator side, a similar limitation is tracked at [OpenAPITools/openapi-generator#21398](https://github.com/OpenAPITools/openapi-generator/issues/21398).  
Until broader `$dynamicRef` support lands across the toolchain, the above pattern has been a practical path for us:

* Keeps schemas DRY on the server side (one envelope, many payloads)
* Keeps generated clients small and type-safe
* Requires only a **tiny** template overlay client-side

### Suggestion (small, opt‑in improvement)

If generator/tooling could *optionally* respect a convention such as:

* `x-api-wrapper: true`
* `x-api-wrapper-datatype: <T>`

…then Java (and other) generators could natively emit thin generic wrappers **without** users shipping custom templates. A generator flag like `--additional-properties=useGenericWrappers=true` would make this opt‑in.

### References (runnable)

* Repo (reference implementation — service + client + templates):

  * [https://github.com/bsayli/spring-boot-openapi-generics-clients](https://github.com/bsayli/spring-boot-openapi-generics-clients)
* Write‑up (why + how):

  * [https://medium.com/@baris.sayli/type-safe-generic-api-responses-with-spring-boot-3-4-openapi-generator-and-custom-templates-ccd93405fb04](https://medium.com/@baris.sayli/type-safe-generic-api-responses-with-spring-boot-3-4-openapi-generator-and-custom-templates-ccd93405fb04)

Happy to open a PR (template partial + conditional) or contribute docs if this direction sounds useful.  
Looking forward to your feedback!

**Comment by handrews** [comment]
It would be a lot better to support the actual intended feature that solves this problem than to further fracture the landscape by implementing alternatives.

**Comment by bsayli** [comment]
@handrews
Totally agree — proper $dynamicRef / $dynamicAnchor support is the right long-term solution, no doubt.

The approach I shared isn’t meant to replace that; it’s a pragmatic bridge for teams using Springdoc + Swagger-core + OpenAPI Generator today, where native support is still missing.

Given that this issue has been open since 2019, many production teams have had to find workable patterns to stay DRY and type-safe in the meantime.

The goal here is to contribute a proven, interim solution — one that:

keeps schemas DRY on the producer side,

generates type-safe clients without duplication,

and requires only minimal customization.


I’d be happy to collaborate or open a PR if this direction can help shape the eventual built-in support. It’s shared in the spirit of contribution, not fragmentation.

**Comment by jochenberger** [comment]
See also #2916.

**Comment by jochenberger** [comment]
and https://github.com/swagger-api/swagger-core/issues/4440
